lib/space_saving.c

```diff
@@ -0,0 +1,795 @@
+					malloc(sizeof(struct counter_bucket));
+
+	ret = add_bucket_id(ct_table, flow, &bkt_id);
+	if(ret < 0)
+		goto out;
+
+	//avl_update(bkt_id, 1);
+
+	ret = get_bucket(bkt_id, flow->proto, ct_bkt);
+	if(ret < 0)
+		goto out;
+
+	ret = add_bucket_entry(ct_bkt, flow, &(data->bkt_data));
+	if(ret < 0)
+		goto out;
+
+	return ret;
+
+out:
+	ret = -1;
+	return ret;
+}
+
+/* Deletes an IP entry from the IP counter */
+static int
+delete_ip_data(struct counter_table *ct_table, struct ip_flow *flow)
+{
+	int ret;
+	int bkt_id;
+	struct counter_bucket *ct_bkt = (struct counter_bucket *)
+					malloc(sizeof(struct counter_bucket));
+
+	ret = get_bucket_id(ct_table, flow, &bkt_id);
+	if(ret < 0)
+		goto out;
+
+	//avl_update(bkt_id, -1);
+
+	ret = get_bucket(bkt_id, flow->proto, ct_bkt);
+	if(ret < 0)
+		goto out;
+
+	ret = delete_bucket_id(ct_table, flow);
+	if(ret < 0)
+		goto out;
+
+	ret = delete_bucket_entry(ct_bkt, flow);
+	if(ret < 0) 
+		goto out;
+
+	return ret;
+
+out:
+	ret = -1;
+	return ret;
+}
+
+/* Moves the ip entry to next higher bucket */
+static int
+update_ip_data(struct counter_table *ct_table, struct ip_flow *flow,
+		struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int ret;
+	struct ip_data *data = (struct ip_data *)
+				malloc(sizeof(struct ip_data));
+
+	struct counter_bucket *ct_bkt = (struct counter_bucket *)
+					malloc(sizeof(struct counter_bucket));
+
+	ret = get_ip_data(ct_table, flow, data);
+	if(ret < 0)
+		goto out;
+
+	ret = update_bucket_id(ct_table, flow);
+	if(ret < 0)
+		goto out;
+
+	//avl_update(data->bkt_id, -1);
+	data->bkt_id++;
+	//avl_update(data->bkt_id, 1);
+
+	ret = delete_ip_data(ct_table, flow);
+	if(ret < 0)
+		goto out;
+
+	ret = get_bucket(data->bkt_id, flow->proto, ct_bkt);
+
+	/* If next bucket is not found, create a new bucket */
+	if(ret == -2) {
+		ret = create_bucket(iface, gk_conf, data->bkt_id);
+		/*
+		 * if(ret < 0)
+		 *       goto out;
+		 */
+		ret = get_bucket(data->bkt_id, flow->proto, ct_bkt);
+		if(ret < 0)
+			goto out;
+	}
+
+	ret = add_ip_data(ct_table, flow, data);
+	if(ret < 0)
+		goto out;
+
+	return ret;
+
+out:
+	ret = -1;
+	return ret;
+}
+
+/*
+static int
+iterate_counter_table(struct counter_bkt *ct_bkt, int proto, int threshold,
+			struct bucket_data *data)
+{
+	uint32_t ret;
+	int count = 0;
+	struct rte_flow *flow;
+	int bkt_id;
+	int i;
+	uint32_t next = 0;
+
+	switch(proto) {
+	case ETHER_TYPE_IPv4:
+		{
+		        time_t curr_time = time(0);
+		        if(curr_time - start_time > 2)
+		                break;
+		        
+		        ret = rte_hash_iterate(counter->ct_ip4,
+		                               (const void **)&flow,
+		                               (void **)&bkt_id, &next);
+		        
+		}
+		break;
+
+	case ETHER_TYPE_IPv6:
+		break;
+
+	default:
+		RTE_LOG(ERR< HASH, "Unknown network layer protocol!\n");
+		ret = -1;
+		break;
+	}
+	return ret;
+}
+*/
+
+/* Bucket id of element with minimum frequency */
+int min_bkt_id;
```

Comment:
Must it be a global variable? Also, please move it to the beginning of the file if it's necessary.